Address Safer CPP warnings in NetworkCacheSpeculativeManager

csaavedra opened the pull request:
#### c996ce015f8ea95665fe58b743e319cb639d7891
<pre>
Address Safer CPP warnings in NetworkCacheSpeculativeManager
<a href="https://bugs.webkit.org/show_bug.cgi?id=292985">https://bugs.webkit.org/show_bug.cgi?id=292985</a>

Reviewed by Chris Dumez.

* Source/WebKit/NetworkProcess/cache/NetworkCacheSpeculativeLoadManager.cpp:
(WebKit::NetworkCache::SpeculativeLoadManager::canRetrieve const):
(WebKit::NetworkCache::SpeculativeLoadManager::registerLoad):
(WebKit::NetworkCache::SpeculativeLoadManager::addPreloadedEntry):
(WebKit::NetworkCache::SpeculativeLoadManager::preconnectForSubresource):
(WebKit::NetworkCache::SpeculativeLoadManager::revalidateSubresource):
(WebKit::NetworkCache::SpeculativeLoadManager::preloadEntry):
(WebKit::NetworkCache::SpeculativeLoadManager::startSpeculativeRevalidation):
* Source/WebKit/SaferCPPExpectations/UncheckedLocalVarsCheckerExpectations:
* Source/WebKit/SaferCPPExpectations/UncountedLambdaCapturesCheckerExpectations:

Canonical link: <a href="https://commits.webkit.org/294972@main">https://commits.webkit.org/294972@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d50a431c2e0c6bc7763e51a656a0034475a62a03

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/103644 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/23327 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/13647 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/108820 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/54292 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/23684 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/31877 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/78761 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/106650 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/18378 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/93501 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/59096 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/18189 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/53645 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/87971 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/11607 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/111198 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/30792 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/22712 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/87754 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/31152 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/89701 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/87401 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/22256 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/32281 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/10005 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/25149 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/30719 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/36031 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/30519 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/33850 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/32080 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebKit/NetworkProcess/cache/NetworkCacheSpeculativeLoadManager.cpp:
```diff
@@ -379,7 +379,10 @@ void SpeculativeLoadManager::registerLoad(GlobalFrameID frameID, const ResourceR
 
         // Start tracking loads in this frame.
-        auto pendingFrameLoad = PendingFrameLoad::create(protectedStorage(), resourceKey, [this, frameID] {
+        auto pendingFrameLoad = PendingFrameLoad::create(protectedStorage(), resourceKey, [this, weakThis = WeakPtr { *this }, frameID] {
```
I doubt this silences the warnings if you capture `this` and `weakThis` here. I suspect you'll have to stop capturing `this` and convert the `weakThis` to a `checkedThis` inside the lambda.

csaavedra replied:
You're right, I'm not sure what was I thinking here :)

cdumez commented on Source/WebKit/NetworkProcess/cache/NetworkCacheSpeculativeLoadManager.cpp:
```diff
@@ -379,19 +379,22 @@ void SpeculativeLoadManager::registerLoad(GlobalFrameID frameID, const ResourceR
+        auto pendingFrameLoad = PendingFrameLoad::create(protectedStorage(), resourceKey, [weakThis = WeakPtr { *this }, frameID] {
+            if (!weakThis)
+                return;
+            CheckedPtr checkedThis = weakThis.get();
```
We normally convert the WeakPtr to a non-weak one *and then* null check that new pointer, rather than null checking the WeakPtr first. It doesn't make a difference in single-threaded code but it is still good practice.

csaavedra replied:
My bad, fixed now.

cdumez commented on Source/WebKit/NetworkProcess/cache/NetworkCacheSpeculativeLoadManager.cpp:
```diff
@@ -379,19 +379,22 @@ void SpeculativeLoadManager::registerLoad(GlobalFrameID frameID, const ResourceR
         // Retrieve the subresources entry if it exists to start speculative revalidation and to update it.
-        retrieveSubresourcesEntry(resourceKey, [this, weakThis = WeakPtr { *this }, frameID, pendingFrameLoad = WTFMove(pendingFrameLoad), requestIsAppInitiated = request.isAppInitiated(), isNavigatingToAppBoundDomain, allowPrivacyProxy, advancedPrivacyProtections](std::unique_ptr<SubresourcesEntry> entry) {
+        retrieveSubresourcesEntry(resourceKey, [weakThis = WeakPtr { *this }, frameID, pendingFrameLoad = WTFMove(pendingFrameLoad), requestIsAppInitiated = request.isAppInitiated(), isNavigatingToAppBoundDomain, allowPrivacyProxy, advancedPrivacyProtections](std::unique_ptr<SubresourcesEntry> entry) {
             if (!weakThis)
```
ditto.

cdumez commented on Source/WebKit/NetworkProcess/cache/NetworkCacheSpeculativeLoadManager.cpp:
```diff
@@ -419,13 +422,16 @@ void SpeculativeLoadManager::addPreloadedEntry(std::unique_ptr<Entry> entry, con
-    m_preloadedEntries.add(key, makeUnique<PreloadedEntry>(WTFMove(entry), WTFMove(revalidationRequest), [this, key, frameID] {
-        auto preloadedEntry = m_preloadedEntries.take(key);
+    m_preloadedEntries.add(key, makeUnique<PreloadedEntry>(WTFMove(entry), WTFMove(revalidationRequest), [weakThis = WeakPtr { *this }, key, frameID] {
+        if (!weakThis)
```
ditto.

cdumez commented on Source/WebKit/NetworkProcess/cache/NetworkCacheSpeculativeLoadManager.cpp:
```diff
@@ -512,11 +518,12 @@ void SpeculativeLoadManager::revalidateSubresource(const SubresourceInfo& subres
-        pendingLoad->addPostMainResourceResponseTask([this, subresourceInfo, entry = WTFMove(entry), frameID, isNavigatingToAppBoundDomain, allowPrivacyProxy, advancedPrivacyProtections]() mutable {
-            if (m_pendingPreloads.contains(subresourceInfo.key()))
+        pendingLoad->addPostMainResourceResponseTask([weakThis = WeakPtr { *this }, subresourceInfo, entry = WTFMove(entry), frameID, isNavigatingToAppBoundDomain, allowPrivacyProxy, advancedPrivacyProtections]() mutable {
+            if (!weakThis || weakThis->m_pendingPreloads.contains(subresourceInfo.key()))
```
Ditto.

cdumez commented on Source/WebKit/NetworkProcess/cache/NetworkCacheSpeculativeLoadManager.cpp:
```diff
@@ -525,21 +532,23 @@ void SpeculativeLoadManager::revalidateSubresource(const SubresourceInfo& subres
         ASSERT(!revalidatedEntry || revalidatedEntry->key() == key);
-
-        auto protectRevalidator = m_pendingPreloads.take(key);
+        if (!weakThis)
```
ditto.

cdumez commented on Source/WebKit/NetworkProcess/cache/NetworkCacheSpeculativeLoadManager.cpp:
```diff
@@ -587,27 +596,27 @@ void SpeculativeLoadManager::preloadEntry(const Key& key, const SubresourceInfo&
     
-    retrieveEntryFromStorage(subresourceInfo, [this, weakThis = WeakPtr { *this }, key, subresourceInfo, frameID, isNavigatingToAppBoundDomain, allowPrivacyProxy, advancedPrivacyProtections](std::unique_ptr<Entry> entry) {
+    retrieveEntryFromStorage(subresourceInfo, [weakThis = WeakPtr { *this }, key, subresourceInfo, frameID, isNavigatingToAppBoundDomain, allowPrivacyProxy, advancedPrivacyProtections](std::unique_ptr<Entry> entry) {
         if (!weakThis)
```
ditto.

cdumez commented on Source/WebKit/NetworkProcess/cache/NetworkCacheSpeculativeLoadManager.cpp:
```diff
@@ -620,9 +629,12 @@ void SpeculativeLoadManager::startSpeculativeRevalidation(const GlobalFrameID& f
-                logSpeculativeLoadingDiagnosticMessage(protectedCache()->protectedNetworkProcess(), frameID, DiagnosticLoggingKeys::entryRightlyNotWarmedUpKey());
-                m_notPreloadedEntries.remove(key);
+            m_notPreloadedEntries.add(key, makeUnique<ExpiringEntry>([weakThis = WeakPtr { *this }, key, frameID] {
+                if (!weakThis)
```
ditto.

cdumez reviewed (approved):
LGTM with fixes.

csaavedra committed:
Address Safer CPP warnings in NetworkCacheSpeculativeManager
https://bugs.webkit.org/show_bug.cgi?id=292985

Reviewed by Chris Dumez.

* Source/WebKit/NetworkProcess/cache/NetworkCacheSpeculativeLoadManager.cpp:
(WebKit::NetworkCache::SpeculativeLoadManager::canRetrieve const):
(WebKit::NetworkCache::SpeculativeLoadManager::registerLoad):
(WebKit::NetworkCache::SpeculativeLoadManager::addPreloadedEntry):
(WebKit::NetworkCache::SpeculativeLoadManager::preconnectForSubresource):
(WebKit::NetworkCache::SpeculativeLoadManager::revalidateSubresource):
(WebKit::NetworkCache::SpeculativeLoadManager::preloadEntry):
(WebKit::NetworkCache::SpeculativeLoadManager::startSpeculativeRevalidation):
* Source/WebKit/SaferCPPExpectations/UncheckedLocalVarsCheckerExpectations:
* Source/WebKit/SaferCPPExpectations/UncountedLambdaCapturesCheckerExpectations:

Canonical link: https://commits.webkit.org/294972@main
